src/core/shape_inference/include/sparse_fill_empty_rows_shape_inference.hpp

```diff
@@ -0,0 +1,125 @@
+                           "The values input must be a 1D tensor. Got: ",
+                           values_shape);
+
+    const auto& dense_shape = input_shapes[1];
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           dense_shape.rank().compatible(1),
+                           "The dense_shape input must be a 1D tensor. Got: ",
+                           dense_shape);
+    if (dense_shape.rank().is_static()) {
+        NODE_SHAPE_INFER_CHECK(
+            op,
+            input_shapes,
+            dense_shape[0].compatible(2),
+            "The dense_shape input must have exactly 2 elements. Only 2D sparse tensors are supported.");
+    }
+
+    const auto& indices_shape = input_shapes[2];
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           indices_shape.rank().compatible(2),
+                           "The indices input must be a 2D tensor. Got: ",
+                           indices_shape);
+
+    if (indices_shape.rank().is_static()) {
+        NODE_SHAPE_INFER_CHECK(
+            op,
+            input_shapes,
+            indices_shape[1].compatible(2),
+            "The indices_shape's second dimension must have 2 elements. Only 2D sparse tensors are supported.");
+        if (values_shape.rank().is_static()) {
+            NODE_SHAPE_INFER_CHECK(op,
+                                   input_shapes,
+                                   indices_shape[0].compatible(values_shape[0]),
+                                   "The first dimension of indices must match the size of values.");
+        }
+    }
+
+    const auto& default_value_shape = input_shapes[3];
+    NODE_SHAPE_INFER_CHECK(op,
+                           input_shapes,
+                           default_value_shape.rank().compatible(0),
+                           "The default_value input must be a scalar. Got: ",
+                           default_value_shape);
+
+    auto output_shapes = std::vector<TRShape>(3);
+    auto& output_indices_shape = output_shapes[0];
+    auto& output_values_shape = output_shapes[1];
+    auto& empty_row_indicator_shape = output_shapes[2];
+    output_indices_shape.resize(2);
+    output_values_shape.resize(1);
+    empty_row_indicator_shape.resize(1);
+    output_indices_shape[1] = 2;  // Only 2D cases are supported
+
+    if (auto dense_shape_value = get_input_const_data_as_shape<TRShape>(op, 1, tensor_accessor)) {
+        const auto& number_of_rows = (*dense_shape_value)[0].get_length();
+        empty_row_indicator_shape[0] = number_of_rows;
+
+        if (auto indices_value = get_input_const_data_as<TRShape, int64_t>(op, 2, tensor_accessor)) {
+            // Rows can be referenced multiple times in sparse representation
+            std::unordered_set<int64_t> existing_rows;
+            const auto& indices_data = *indices_value;
+            size_t indices_count = indices_data.size() / 2;
```

Comment:
The logic to calculate the number of empty rows seems to be also an important part of the actual op logic. 
Originally I thought it will always produce dynamic dimensions (similar to ops like NonZero), but as it's possible to infer those values at the shape_infer stage, it can be done this way, but worth to consider (possibly as further improvement) some helpers for common code between reference and shape_infer.

Reply:
Probably a good idea, let's move this to https://github.com/openvinotoolkit/openvino/pull/30307 where it can actually be applied fully. I'll check if I can move some logic to common code.